File: server/src/FloVMP.Core/Native/UdpProtocol.cs
using System.Buffers.Binary;
using System.Net;
using System.Security.Cryptography;
using System.Text;

namespace FloVMP.Core.Native;

/// <summary>FLOV/3 wire format. This is a codec only; it does not open a UDP socket.</summary>
public static class UdpProtocol
{
    public const byte Version = 3;
    public const int HeaderSize = 32;
    public const int MaxDatagramBytes = 1200;
    public const int MaxPayloadBytes = MaxDatagramBytes - HeaderSize;

    public enum PacketType : byte
    {
        Hello = 1, Cookie = 2, Connect = 3, Challenge = 4,
        Auth = 5, Accept = 6, Reliable = 7, State = 8, Ack = 9,
    }

    public readonly record struct Header(PacketType Type, ulong ConnectionId,
        uint Sequence, uint Ack, uint AckBits, ushort PayloadLength);

    public static bool TryRead(ReadOnlySpan<byte> packet, out Header header, out ReadOnlySpan<byte> payload)
    {
        header = default;
        payload = default;
        if (packet.Length is < HeaderSize or > MaxDatagramBytes ||
            !packet[..4].SequenceEqual("FLV3"u8) || packet[4] != Version ||
            packet[5] is < 1 or > 9 || packet[6] != 0 || packet[7] != 0 ||
            packet[30] != 0 || packet[31] != 0) return false;

        var length = BinaryPrimitives.ReadUInt16BigEndian(packet[28..]);
        if (length > MaxPayloadBytes || packet.Length != HeaderSize + length) return false;
        header = new Header((PacketType)packet[5],
            BinaryPrimitives.ReadUInt64BigEndian(packet[8..]),
            BinaryPrimitives.ReadUInt32BigEndian(packet[16..]),
            BinaryPrimitives.ReadUInt32BigEndian(packet[20..]),
            BinaryPrimitives.ReadUInt32BigEndian(packet[24..]), length);
        payload = packet[HeaderSize..];
        return true;
    }

    public static bool TryWrite(Span<byte> destination, PacketType type, ulong connectionId,
        uint sequence, uint ack, uint ackBits, ReadOnlySpan<byte> payload, out int written)
    {
        written = 0;
        if ((byte)type is < 1 or > 9 || payload.Length > MaxPayloadBytes ||
            destination.Length < HeaderSize + payload.Length) return false;
        var packet = destination[..(HeaderSize + payload.Length)];
        packet[..HeaderSize].Clear();
        "FLV3"u8.CopyTo(packet);
        packet[4] = Version;
        packet[5] = (byte)type;
        BinaryPrimitives.WriteUInt64BigEndian(packet[8..], connectionId);
        BinaryPrimitives.WriteUInt32BigEndian(packet[16..], sequence);
        BinaryPrimitives.WriteUInt32BigEndian(packet[20..], ack);
        BinaryPrimitives.WriteUInt32BigEndian(packet[24..], ackBits);
        BinaryPrimitives.WriteUInt16BigEndian(packet[28..], (ushort)payload.Length);
        payload.CopyTo(packet[HeaderSize..]);
        written = packet.Length;
        return true;
    }
}

/// <summary>
/// Stateless address proof for the first two packets. A cookie is valid only
/// for the same IP, UDP port, client nonce and the current/previous 10s bucket.
/// It is not an authentication token: the existing client-key signature still
/// has to be checked before the game session is created.
/// </summary>
public sealed class UdpCookieProtector
{
    public const int NonceSize = 16;
    public const int CookieSize = 24; // 8-byte time bucket + 16-byte HMAC tag
    public const int BucketSeconds = 10;
    private static readonly byte[] Domain = Encoding.ASCII.GetBytes("FLOVMP-UDP-COOKIE-v3");
    private readonly byte[] _secret;

    public UdpCookieProtector(ReadOnlySpan<byte> secret)
    {
        if (secret.Length != 32) throw new ArgumentException("cookie secret must be 32 bytes", nameof(secret));
        _secret = secret.ToArray();
    }

    public static UdpCookieProtector CreateRandom() => new(RandomNumberGenerator.GetBytes(32));

    public bool TryCreate(IPAddress address, ushort port, ReadOnlySpan<byte> nonce,
        DateTimeOffset now, Span<byte> cookie)
    {
        if (nonce.Length != NonceSize || cookie.Length < CookieSize) return false;
        var bucket = now.ToUnixTimeSeconds() / BucketSeconds;
        BinaryPrimitives.WriteInt64BigEndian(cookie, bucket);
        ComputeTag(address, port, nonce, bucket, cookie.Slice(8, 16));
        return true;
    }

    public bool Validate(IPAddress address, ushort port, ReadOnlySpan<byte> nonce,
        ReadOnlySpan<byte> cookie, DateTimeOffset now)
    {
        if (nonce.Length != NonceSize || cookie.Length != CookieSize) return false;
        var bucket = BinaryPrimitives.ReadInt64BigEndian(cookie);
        var current = now.ToUnixTimeSeconds() / BucketSeconds;
        if (bucket != current && bucket != current - 1) return false;
        Span<byte> expected = stackalloc byte[16];
        ComputeTag(address, port, nonce, bucket, expected);
        return CryptographicOperations.FixedTimeEquals(expected, cookie[8..]);
    }

    private void ComputeTag(IPAddress address, ushort port, ReadOnlySpan<byte> nonce,
        long bucket, Span<byte> tag)
    {
        Span<byte> input = stackalloc byte[Domain.Length + 1 + 16 + 2 + NonceSize + 8];
        Domain.CopyTo(input);
        var offset = Domain.Length;
        input[offset++] = UdpProtocol.Version;
        address.MapToIPv6().TryWriteBytes(input[offset..], out _);
        offset += 16;
        BinaryPrimitives.WriteUInt16BigEndian(input[offset..], port);
        offset += 2;
        nonce.CopyTo(input[offset..]);
        offset += NonceSize;
        BinaryPrimitives.WriteInt64BigEndian(input[offset..], bucket);
        Span<byte> hash = stackalloc byte[32];
        HMACSHA256.HashData(_secret, input, hash);
        hash[..tag.Length].CopyTo(tag);
        CryptographicOperations.ZeroMemory(hash);
    }
}

File: native/legacy-3889/client/tools/udp_protocol.py
"""FLOV/3 wire codec for the Python test bot. No UDP game transport yet."""

import hashlib
import hmac
import ipaddress
import struct

MAGIC = b"FLV3"
VERSION = 3
HEADER = struct.Struct(">4sBBBBQIIIHH")
HEADER_SIZE = HEADER.size
MAX_DATAGRAM = 1200
MAX_PAYLOAD = MAX_DATAGRAM - HEADER_SIZE
COOKIE_SIZE = 24
NONCE_SIZE = 16
COOKIE_DOMAIN = b"FLOVMP-UDP-COOKIE-v3"
PACKET_TYPES = set(range(1, 10))


def pack_packet(kind, connection_id=0, sequence=0, ack=0, ack_bits=0, payload=b""):
    if kind not in PACKET_TYPES or len(payload) > MAX_PAYLOAD:
        raise ValueError("invalid type or payload length")
    return HEADER.pack(MAGIC, VERSION, kind, 0, 0, connection_id,
                       sequence, ack, ack_bits, len(payload), 0) + payload


def unpack_packet(data):
    if not HEADER_SIZE <= len(data) <= MAX_DATAGRAM:
        raise ValueError("invalid datagram length")
    magic, version, kind, flags, reserved, cid, seq, ack, ack_bits, size, tail = HEADER.unpack_from(data)
    if (magic != MAGIC or version != VERSION or kind not in PACKET_TYPES or
            flags or reserved or tail or size > MAX_PAYLOAD or len(data) != HEADER_SIZE + size):
        raise ValueError("invalid header")
    return (kind, cid, seq, ack, ack_bits, data[HEADER_SIZE:])


def _mapped_ip(address):
    ip = ipaddress.ip_address(address)
    if isinstance(ip, ipaddress.IPv4Address):
        ip = ipaddress.IPv6Address("::ffff:" + str(ip))
    return ip.packed


def make_cookie(secret, address, port, nonce, unix_seconds):
    if len(secret) != 32 or len(nonce) != NONCE_SIZE or not 0 <= port <= 65535:
        raise ValueError("invalid cookie input")
    bucket = unix_seconds // 10
    msg = (COOKIE_DOMAIN + bytes([VERSION]) + _mapped_ip(address) +
           struct.pack(">H", port) + nonce + struct.pack(">q", bucket))
    return struct.pack(">q", bucket) + hmac.new(secret, msg, hashlib.sha256).digest()[:16]


def check_cookie(secret, address, port, nonce, cookie, unix_seconds):
    if len(cookie) != COOKIE_SIZE:
        return False
    bucket = struct.unpack_from(">q", cookie)[0]
    if bucket not in (unix_seconds // 10, unix_seconds // 10 - 1):
        return False
    return hmac.compare_digest(make_cookie(secret, address, port, nonce, bucket * 10), cookie)
